### PR TITLE
Add GCS Bucket Mirrors

### PR DESCRIPTION
**Pull request by douglasjacobsen**
This pull request contains changes to support Google Cloud Storage buckets as mirrors, meant for hosting Spack build-caches. This feature is beneficial for folks that are running infrastructure on Google Cloud Platform. On public cloud systems, resources are ephemeral and in many cases, installing compilers, MPI flavors, and user packages from scratch takes up considerable time.

Giving users the ability to host a Spack mirror that can store build caches in GCS buckets offers a clean solution for reducing application rebuilds for Google Cloud infrastructure.

### CODE REVIEW

**Comment by douglasjacobsen** [comment]
This PR supersedes https://github.com/spack/spack/pull/24422

File: lib/spack/spack/fetch_strategy.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
Comment (tgamblin) [inline lib/spack/spack/fetch_strategy.py]: 2021

File: var/spack/repos/builtin/packages/lua/package.py [hunk 2]
```diff
@@ -66,7 +66,7 @@ def install(self, spec, prefix):
              'MYLDFLAGS=-L%s -L%s' % (
                  spec['readline'].prefix.lib,
                  spec['ncurses'].prefix.lib),
-             'MYLIBS=-lncursesw -ltinfow',
+             'MYLIBS=-lncursesw -ltinfo',
```
Comment (douglasjacobsen) [inline var/spack/repos/builtin/packages/lua/package.py]: This change is unrelated and should probably be removed.

File: lib/spack/spack/util/web.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
Comment (douglasjacobsen) [inline lib/spack/spack/util/web.py]: 2021 as well.

File: lib/spack/spack/util/gcs.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
Comment (douglasjacobsen) [inline lib/spack/spack/util/gcs.py]: 2021 here too.

**Review by scottwittenburg — COMMENTED** [review]
This looks pretty good to me, but since it changes some things with respect to how S3 is handled, it's probably a good idea for @opadron to take a look.

**Review by scottwittenburg — CHANGES_REQUESTED** [review]
I just noticed some cleanup jobs failing in gitlab pipelines, to do with the `recursive` argument to `remove_url()`.  Here's a link to a failing job, though you can find it from the gitlab check on your PR as well.

https://gitlab.spack.io/spack/spack/-/jobs/1080302

**Comment by opadron** [comment]
Why do we need this?  Isn't GCS S3-compatible?

**Comment by douglasjacobsen** [comment]
This allows us to use google-cloud-storage's python API without having a dependence on boto3.

However, users could choose to use the s3 backend, and configure it to point to GCS resources by changing the environment variables instead of using this native GCS backend.

**Comment by scottwittenburg** [comment]
> Why do we need this? Isn't GCS S3-compatible?

Looks like it's adding a new url scheme, so I assumed GCS could be addressed with both url protocols.

File: lib/spack/spack/util/web.py [hunk 5]
```diff
@@ -209,14 +223,19 @@ def url_exists(url):
 
     if url.scheme == 's3':
         s3 = s3_util.create_s3_session(url)
+        from botocore.exceptions import ClientError
         try:
-            s3.get_object(Bucket=url.netloc, Key=url.path.lstrip('/'))
```
Comment (opadron) [inline lib/spack/spack/util/web.py]: Why change this?  Is the need for the `lstrip()` handled elsewhere, and I'm not seeing it?

**Comment by opadron** [comment]
> This allows us to use google-cloud-storage's python API without having a dependence on boto3.

Ok, but to do that requires a dependency on the google-cloud-storage's python API.  I'm not yet convinced this is worth adding on the basis of reducing dependencies because it doesn't.

Is there some kind of feature or capability that GCS provides that S3 does not?  And, are we currently or planning to take advantage of those capabilities?  Otherwise, I'm having a really hard time seeing why these changes are needed.

Maybe my reading of the above quote is off.  Are you perhaps saying that using boto3 with GCS would *also* require the google-cloud-storage's python API?

**Comment by scottwittenburg** [comment]
> I just noticed some cleanup jobs failing in gitlab pipelines, to do with the `recursive` argument to `remove_url()`. Here's a link to a failing job, though you can find it from the gitlab check on your PR as well.

I think the cleanup job uses `mirror.destroy`, which invokes the `remove_url()` method.  Maybe a quick grep for `remove_url` in the repo would find any other call sites.  

But @douglasjacobsen I'm curious if you saw some documentation indicating that we no longer need the special handling for recursive removal of a url?  I would be happy to be enlightened if you could share a link.  Thanks!

File: lib/spack/spack/util/web.py [hunk 6]
```diff
@@ -226,57 +245,22 @@ def url_exists(url):
...
-    if 'Deleted' in result:
-        for d in result['Deleted']:
-            tty.debug('Deleted {0}'.format(d['Key']))
-    if 'Errors' in result:
-        for e in result['Errors']:
-            tty.debug('Failed to delete {0} ({1})'.format(
-                e['Key'], e['Message']))
-
-
-def remove_url(url, recursive=False):
```
Comment (opadron) [inline lib/spack/spack/util/web.py]: As @scottwittenburg pointed out, that `recursive` argument (and the corresponding logic) is needed.  What is the rationale for removing it?

**Comment by douglasjacobsen** [comment]
Yes, using the `gs://` URL scheme would require using the google-cloud-storage API rather than the boto3 API. Using boto3 with GCS does not require the google-cloud-storage API.

These changes more allow users who are already using the Google Cloud SDK to access GCS buckets using their SDK credentials and the default endpoint, similar to how the default endpoint for `s3://` is Amazon's s3. So, the goal of this PR is primarily to ease the barrier of entry for people using google tools to access GCS buckets, and provide an implementation that mirrors what the s3 backend currently does (relying on Amazon tools and libraries).

**Comment by douglasjacobsen** [comment]
@scottwittenburg: Sorry, this is a bug. I'm going to work on fixing it asap. I think the version we started working on didn't have the recusrive destruction of these URLs, and when I rebased these changes didn't translate properly.

For the GCS backend, we could use `force` to try and make the API do a recursive destruction, however that fails if there are more than 256 objects in the bucket. So it's likely safer to iterate similar to how the s3 backend works.

File: lib/spack/spack/util/web.py [hunk 7]
```diff
@@ -592,10 +584,7 @@ def find_versions_of_archive(
         #   .sha256
         #   .sig
         # However, SourceForge downloads still need to end in '/download'.
-        url_regex += r'(\/download)?'
-        # PyPI adds #sha256=... to the end of the URL
-        url_regex += '(#sha256=.*)?'
-        url_regex += '$'
```
Comment (opadron) [inline lib/spack/spack/util/web.py]: I can't really comment on the specifics of these few lines, since I'm not as familiar with their purpose.

But, to me, this appears to be another unrelated change that is being proposed for no obvious reason.  If there's a rationale for changes like these, I'd be happy to consider them.

**Comment by douglasjacobsen** [comment]
We should probably resolve the discussion about whether or not this is useful before I put more time into fixing the PR though. :)

**Review by opadron — CHANGES_REQUESTED** [review]
I'm quite confused by these proposed changes.  They concern me because I worry that it might be unnecessarily adding new code paths to access GCS, which as far as I understand, are already accessible via an `s3://` URL and boto.

Besides the above concern, I notice that a number of the proposed changes modify how `s3://` urls are handled as well as other minor details of how Spack operates without a clear reason.  The PR author themselves notes that one of these changes are unrelated and should be removed.  Perhaps this is supposed to be a WIP and not yet ready for review?

@douglasjacobsen if I am in error, please accept my apologies.  Help us to understand your intent and what we can do to help.

**Comment by opadron** [comment]
@douglasjacobsen so, I just noticed your comments from the last 15 minutes or so.  Thanks for helping to clear up some of my questions.

I think a big part of what you're going for is a good developer experience for users of GCS tooling, is that right?  I can see how it would be more convenient to just use existing tooling, but I think we should balance that desire against the maintenance cost of maintaining `gcs` and `s3` url handling vs just `s3` urls.  What does the experience look like if someone using google cloud sdk needs to install boto and access GCS through our `s3` url handling?  Is there a lot of extra steps that the user would need to take?  Other barriers to use?  These are perhaps some things we should consider.

**Comment by douglasjacobsen** [comment]
@opadron Thanks for the discussion. And I should probably apologize for the changes that are irrelevant. This should not change the s3 backend, and it should not modify how spack works (i.e. the recursive removal is an accident).

This is a branch that was started a long time ago (https://github.com/spack/spack/pull/24422) and I've been working with the original author to fix some of the issues we had with it. The unrelated changes that are in the PR are simply an oversight on my part when I rebased the branch to fix some of the merge conflicts.

Regarding user experience, yes that's exactly what I'm going for. The steps a user needs to go through to access a GCS bucket with the existing s3 backend are something similar to the following (which honestly are not very well documented and it takes a lot for a naive user to figure out):
1. Install boto3 (obvious if you're using an s3 bucket, but seems awkward if you're already using google tooling)
2. Set the `S3_ENDPOINT_URL` env-var to `https://storage.googleapis.com/download/storage/v1` (not very well documented anywhere)
3. Configure Boto to access GCS resources (something like this: https://cloud.google.com/storage/docs/boto-gsutil)

Whereas with this PR, the steps are similar to the user experience with an `s3://` hosted cache/mirror.
1. Install google-cloud-storage
2. Authenticate with `gcloud auth application-default login`

Then you can read from `gs://` based caches and mirrors.

**Comment by opadron** [comment]
I agree that a better user experience could be a compelling argument in favor of changes like these.

However, the best trade off between this interest and maintainability is not clear.  It'd be great if we could have other Spack devs & users weigh in, especially if they're also GCS users.

@tgamblin @adamjstewart @alalazo @gartung @scheibelp: would love to hear your opinion.

Also, @eugeneswalker, I believe you are a user of GCS, is that right?  What do you think about having code in Spack for handling `gs://` urls?

**Comment by tgamblin** [comment]
@opadron: we really want Spack to be useable by the different cloud providers using their own tooling, and this is a pretty small PR as far as they go.  I don't think it's a huge maintenance burden to support this, and it's supported in a very similar way to the way we currently support `boto3`.  I think if we *don't* merge this, we're biasing for one particular cloud provider over another, which is not what we want.

If merging stuff like this enables folks at GCP to deliver Spack efficiently to their users, it ultimately helps the project and get us a broader user base, which is what we want.

So I'd say please evaluate this considering the use of the GCP native libs as a bonus (the same way we consider `boto` support a bonus for AWS).

If there is a good way to support *all* of the object store interfaces smoothly without the native libraries, we could talk about that, but I think at this point we want buildcaches to be just as usable on whatever cloud you happen to be on.

**Comment by gartung** [comment]
I have no comment since we don't use this feature.

**Comment by opadron** [comment]
Thanks, I think @tgamblin made some good points for why we want native support for GCS.  @douglasjacobsen, I think we're in agreement that this PR is useful and worth putting more time in to.  Let us know if you have any more questions.

**Comment by douglasjacobsen** [comment]
@tgamblin, @opadron, and @scottwittenburg: This should now be reviewable. I fixed the `remove_url` recursive issue, and now bucket destruction works as expected.

Feel free to let me know if you have any more questions or comments.

Thanks!

**Comment by douglasjacobsen** [comment]
@tgamblin, @opadron, and @scottwittenburg: OK, it seems like I've fixed most (if not all) of the CI checks. There's one still pending, but let me know if you see any other PR changes that you'd like me to make. I think I resolved all of the previous issues already.

File: lib/spack/spack/fetch_strategy.py [hunk 8]
```diff
@@ -1407,6 +1407,54 @@ def fetch(self):
             raise FailedDownloadError(self.url)
 
 
+@fetcher
+class GCSFetchStrategy(URLFetchStrategy):
+    """FetchStrategy that pulls from a GCS bucket."""
+    url_attr = 'gs'
```
Comment (opadron) [inline lib/spack/spack/fetch_strategy.py:1413]: This is really minor, but why `gs` for the scheme?  Is this a common abbreviation for GCS?  Seems that `gcs` would be less likely to collide with something else we might want to support in the future.

File: lib/spack/spack/util/gcs.py [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+
+        self.bucket = None
+
+    def exists(self):
+        if not self.bucket:
+            try:
+                self.bucket = self.client.bucket(self.name)
+            except Exception as ex:
+                tty.error("{0}, Failed check for bucket existence".format(ex))
+                sys.exit(1)
```
Comment (opadron) [inline lib/spack/spack/util/gcs.py]: Is there a more specific class of exception types that could be caught here?

File: lib/spack/spack/util/web.py [hunk 10]
```diff
@@ -358,6 +382,10 @@ def list_url(url, recursive=False):
             key.split('/', 1)[0]
             for key in _iter_s3_prefix(s3, url)))
 
+    elif url.scheme == 'gs':
+        gcs = gcs_util.GCSBucket(url)
+        return gcs.get_all_blobs()
+
```
Comment (opadron) [inline lib/spack/spack/util/web.py]: It looks like this code always returns a recursive listing, regardless of the value of `recursive`, is that right?  Notice how in the S3 case, we emulate local file system semantics by truncating and filtering out prefixes beyond the first `/` character.

File: lib/spack/spack/util/web.py [hunk 11]
```diff
@@ -516,6 +544,9 @@ def _urlopen(req, *args, **kwargs):
     if url_util.parse(url).scheme == 's3':
         import spack.s3_handler
         opener = spack.s3_handler.open
+    elif url_util.parse(url).scheme == 'gs':
+        import spack.gcs_handler
+        return spack.gcs_handler.gcs_open(req)
```
Comment (opadron) [inline lib/spack/spack/util/web.py]: I'm not sure what the `*args` and `**kwargs` could be in this context, but it's probably a good idea to pass those along to your opener.

```suggestion
    elif url_util.parse(url).scheme == 'gs':
        import spack.gcs_handler
        opener = spack.gcs_handler.gcs_open
```

... and then leave the calling to the code immediately following?

**Review by opadron — CHANGES_REQUESTED** [review]
This is looking really good!  Additional feedback below.

File: lib/spack/spack/fetch_strategy.py (hunk 8 above)
Comment (douglasjacobsen) [reply lib/spack/spack/fetch_strategy.py:1413]: Sure, I can name it gcs. I used `gs` because the URLs are formatted as `gs://` rather than `gcs://`. But I'd happily update this `url_attr` field to match.

Comment (opadron) [reply lib/spack/spack/fetch_strategy.py:1413]: So, that's actually what I was asking (sorry if that wasn't clear).

Would it make more sense to use a `gcs://` scheme for the URLs?  Or perhaps `gs://` is more "standard" among GCS users?

Comment (opadron) [reply lib/spack/spack/fetch_strategy.py:1413]: Of course, I just *now* thought to google it and it does look like `gs://` is what is commonly used.  If this is right, then I think we should stick with `gs`. (Sorry for the noise)

Comment (douglasjacobsen) [reply lib/spack/spack/fetch_strategy.py:1413]: No problem at all!

**Comment by douglasjacobsen** [comment]
@opadron I think I fixed all of the issues you spotted, let me know if you see any new ones though.

**Review by opadron — DISMISSED** [review]

**Comment by douglasjacobsen** [comment]
@scottwittenburg I think I've fixed all of the issues you spotted as well. Feel free to let me know if there are any more modifications you'd like to see though.

**Review by scottwittenburg — DISMISSED** [review]
Looks good to me @douglasjacobsen, thanks!

**Comment by fluidnumericsJoe** [comment]
Just wanted to throw my vote of support for this feature - I've been working with @douglasjacobsen 's contributions on my own fork on Google Cloud and have been able to host binary caches on GCS that help considerably reduce build times for Google Compute Engine VM images.

File: lib/spack/spack/test/gcs_fetch.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+import pytest
+
+import spack.config as spack_config
+import spack.fetch_strategy as spack_fs
+import spack.stage as spack_stage
+
+
+@pytest.mark.parametrize('_fetch_method', ['curl', 'urllib'])
+def test_gcsfetchstrategy_sans_url(_fetch_method):
```
Comment (alalazo) [inline lib/spack/spack/test/gcs_fetch.py]: Maybe:
```suggestion
def test_gcsfetchstrategy_without_url(_fetch_method):
```
? But in any case :tada: :fr: :tada:

File: lib/spack/spack/fetch_strategy.py [hunk 13]
```diff
@@ -1407,6 +1407,54 @@ def fetch(self):
             raise FailedDownloadError(self.url)
 
 
+@fetcher
+class GCSFetchStrategy(URLFetchStrategy):
+    """FetchStrategy that pulls from a GCS bucket."""
```
Comment (alalazo) [inline lib/spack/spack/fetch_strategy.py:1412]: This class is very similar to `S3FetchStrategy`. Either here or in a following PR we should gather in a common place (a base class?) the duplicated code.

File: lib/spack/spack/gcs_handler.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import six.moves.urllib.response as urllib_response
+
+import spack.util.url as url_util
+import spack.util.web as web_util
+
+
+def gcs_open(req, *args, **kwargs):
```
Comment (alalazo) [inline lib/spack/spack/gcs_handler.py:12]: Where are the `*args` and `**kwargs` used? I am wondering why is it necessary to have them in the function signature.

Comment (alalazo) [reply lib/spack/spack/gcs_handler.py:12]: It might be also good to add a brief docstring here.

File: lib/spack/spack/test/gcs_fetch.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+
+import pytest
+
+import spack.config as spack_config
+import spack.fetch_strategy as spack_fs
+import spack.stage as spack_stage
```
Comment (alalazo) [inline lib/spack/spack/test/gcs_fetch.py]: Can we use the fully qualified names:
```suggestion
import spack.config
import spack.fetch_strategy
import spack.stage
```
? For the first and lat modules it won't even be more verbose, it's just `.` instead of `_`

File: lib/spack/spack/util/gcs.py [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+
+
+class GCSBucket(object):
+    """GCS Bucket Object
+    Create a wrapper object for a GCS Bucket. Provides methods to wrap spack
+    related tasks, such as destroy.
+    """
+    def __init__(self, url, client=None):
+        """Constructor for GCSBucket objects
+        """
```
Comment (alalazo) [inline lib/spack/spack/util/gcs.py]: Can you document the arguments briefly? Something like:
```
    Args:
        url (str): ...
        client (?): ...
```

File: lib/spack/spack/util/gcs.py [hunk 17]
```diff
@@ -0,0 +1,216 @@
...
+        if url.scheme != 'gs':
+            raise ValueError('Can not create GCS bucket connection with scheme {SCHEME}'
+                             .format(SCHEME=url.scheme))
+        self. url = url
```
Comment (alalazo) [inline lib/spack/spack/util/gcs.py]: We should remove the space:
```suggestion
        self.url = url
```

File: lib/spack/spack/util/gcs.py [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+
+    def exists(self):
+        from google.cloud.exceptions import NotFound
+        if not self.bucket:
+            try:
+                self.bucket = self.client.bucket(self.name)
+            except NotFound as ex:
+                tty.error("{0}, Failed check for bucket existence".format(ex))
+                sys.exit(1)
+        return (self.bucket is not None)
```
Comment (alalazo) [inline lib/spack/spack/util/gcs.py]: Minor, but parenthesis are not needed (and usually not used in Python):
```suggestion
        return self.bucket is not None
```

File: lib/spack/spack/util/gcs.py [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+                      If false, print absolute blob paths (useful for
+                         destruction of bucket)
+        """
+        tty.debug('Getting GCS blobs... Recurse {0} -- Rel: {1}'.format(
+            recursive, relative))
+
+        if relative:
+            converter = self._relative_blob_name
+        else:
+            converter = str
```
Comment (alalazo) [inline lib/spack/spack/util/gcs.py]: Minor, we can rewrite the following lines without the need for an else
```suggestion
        converter = str
        if relative:
            converter = self._relative_blob_name
```

File: lib/spack/spack/util/gcs.py [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+
+        self.url = url
+        if url.scheme != 'gs':
+            raise ValueError('Can not create GCS blob connection with scheme: {SCHEME}'
+                             .format(SCHEME=url.scheme))
+
+        if not client:
+            self.client = gcs_client()
+        else:
+            self.client = client
```
Comment (alalazo) [inline lib/spack/spack/util/gcs.py]: This can be written more concisely as:
```suggestion
        self.client = client or gcs_client()
```

File: lib/spack/spack/util/gcs.py [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+        return self.bucket.get_blob(self.blob_path).open(mode='rb')
+
+    def get_blob_headers(self):
+        blob = self.bucket.get_blob(self.blob_path)
+
+        headers = {}
+        headers['Content-type'] = blob.content_type
+        headers['Content-encoding'] = blob.content_encoding
+        headers['Content-language'] = blob.content_language
+        headers['MD5Hash'] = blob.md5_hash
```
Comment (alalazo) [inline lib/spack/spack/util/gcs.py]: Here we can use a dict literal:
```suggestion
        headers = {
            'Content-type': blob.content_type,
            'Content-encoding': blob.content_encoding,
            'Content-language': blob.content_language,
            'MD5Hash': blob.md5_hash
        }
```

**Review by alalazo — CHANGES_REQUESTED** [review]
Thanks for this PR! I started having a look at the code, let me know what you think of the comments.

File: lib/spack/spack/fetch_strategy.py (hunk 13 above)
Comment (douglasjacobsen) [reply lib/spack/spack/fetch_strategy.py:1412]: I agree. I think it could be nice (especially as Spack considers possibly other cloud service providers) to have a more unified back-end for all of the cloud storage services.

Personally, I thought this could be outside of the scope of this PR, but it would be nice to do it at some point.

File: lib/spack/spack/gcs_handler.py (hunk 14 above)
Comment (douglasjacobsen) [reply lib/spack/spack/gcs_handler.py:12]: I only added `*args` and `**kwargs` to keep the signature the same as the other functions that are used as function pointers here: https://github.com/douglasjacobsen/spack/blob/b9370ba0d7e54f7de37ef4d5f680479aeb741d11/lib/spack/spack/util/web.py#L552

Would you like me to change it so I don't have them in the signature? I'll work on the docstring in the back ground.

File: lib/spack/spack/test/gcs_fetch.py (hunk 15 above)
Comment (douglasjacobsen) [reply lib/spack/spack/test/gcs_fetch.py]: Definitely. :)

File: lib/spack/spack/util/gcs.py (hunk 16 above)
Comment (douglasjacobsen) [reply lib/spack/spack/util/gcs.py]: 👍🏻

File: lib/spack/spack/util/gcs.py (hunk 17 above)
Comment (douglasjacobsen) [reply lib/spack/spack/util/gcs.py]: 👍🏻

File: lib/spack/spack/util/gcs.py (hunk 18 above)
Comment (douglasjacobsen) [reply lib/spack/spack/util/gcs.py]: 👍🏻

File: lib/spack/spack/util/gcs.py (hunk 19 above)
Comment (douglasjacobsen) [reply lib/spack/spack/util/gcs.py]: 👍🏻

File: lib/spack/spack/util/gcs.py (hunk 20 above)
Comment (douglasjacobsen) [reply lib/spack/spack/util/gcs.py]: 👍🏻

File: lib/spack/spack/util/gcs.py (hunk 21 above)
Comment (douglasjacobsen) [reply lib/spack/spack/util/gcs.py]: 👍🏻

File: lib/spack/spack/test/gcs_fetch.py (hunk 12 above)
Comment (douglasjacobsen) [reply lib/spack/spack/test/gcs_fetch.py]: 👍🏻

**Comment by douglasjacobsen** [comment]
@alalazo: Let me know if these changes look good.

Thanks for the review!

File: lib/spack/spack/fetch_strategy.py (hunk 13 above)
Comment (alalazo) [reply lib/spack/spack/fetch_strategy.py:1412]: > Personally, I thought this could be outside of the scope of this PR, but it would be nice to do it at some point.

:+1:

File: lib/spack/spack/util/gcs.py [hunk 22]
```diff
@@ -0,0 +1,218 @@
...
+        self.name = self.url.netloc
+        if self.url.path[0] == '/':
+            self.prefix = self.url.path[1:]
+        else:
+            self.prefix = self.url.path
+
+        if not client:
+            self.client = gcs_client()
+        else:
+            self.client = client
```
Comment (alalazo) [inline lib/spack/spack/util/gcs.py]: Apologies, missed this one in the first review:
```suggestion
        self.client = client or gcs_client()
```

**Review by alalazo — COMMENTED** [review]
This LGTM. There's only a minor comment on a thing I overlooked on first review.

**Comment by douglasjacobsen** [comment]
No problem. Thanks @alalazo

For some reason some style errors came up in the last CI check, but I don't see them on my local machine. We'll see if they crop up again after it runs this time.

**Review by alalazo — APPROVED** [review]

**Comment by alalazo** [comment]
Thanks!
